vpr/src/route/heap_type.h

```diff
@@ -0,0 +1,162 @@
+ * Each heap element represents a partial route.
+ *
+ * cost:    The cost used to sort heap.
+ *          For the timing-driven router this is the backward_path_cost +
+ *          expected cost to the target.
+ *          For the breadth-first router it is the node cost to reach this
+ *          point.
+ *
+ * backward_path_cost:  Used only by the timing-driven router.  The "known"
+ *                      cost of the path up to and including this node.
+ *                      In this case, the .cost member contains not only
+ *                      the known backward cost but also an expected cost
+ *                      to the target.
+ *
+ * R_upstream: Used only by the timing-driven router.  Stores the upstream
+ *             resistance to ground from this node, including the
+ *             resistance of the node itself (device_ctx.rr_nodes[index].R).
+ *
+ * index: The RR node index associated with the costs/R_upstream values
+ *
+ * u.prev.node: The previous node used to reach the current 'index' node
+ * u.prev.next: The edge from u.prev.node used to reach the current 'index' node
+ *
+ * u.next:  pointer to the next s_heap structure in the free
+ *          linked list.  Not used when on the heap.
+ *
+ */
+struct t_heap {
+    float cost = 0.;
+    float backward_path_cost = 0.;
+    float R_upstream = 0.;
+
+    int index = OPEN;
+
+    struct t_prev {
+        int node;
+        int edge;
+    };
+
+    union {
+        t_heap* next;
+        t_prev prev;
+    } u;
+};
+
+// t_heap object pool, useful for implementing heaps that conform to
+// HeapInterface.
+class HeapStorage {
+  public:
+    HeapStorage();
+
+    // Allocate a heap item.
+    t_heap* alloc();
+
+    // Free a heap item.
+    void free(t_heap* hptr);
+    void free_all_memory();
+
+  private:
+    /* For keeping track of the sudo malloc memory for the heap*/
+    vtr::t_chunk heap_ch_;
+
+    t_heap* heap_free_head_;
+    size_t num_heap_allocated_;
+};
+
+// Interface to heap used for router optimization.
+//
+// Note: Objects used in instances of HeapInterface must always be allocated
+// and free'd using the HeapInterface::alloc and HeapInterface::free methods
+// of that instance.  Object pools are likely in use.
+//
+// As a general rule, any t_heap objects returned from this interface,
+// **must** be HeapInterface::free'd before destroying the HeapInterface
+// instance.  This ensure that no leaks are present in the users of the heap.
+// Violating this assumption may result in a assertion violation.
+class HeapInterface {
+  public:
+    virtual ~HeapInterface() {}
+
+    // Allocate a heap item.
+    //
+    // This transfers ownership of the t_heap object from HeapInterface to the
+    // caller.
+    virtual t_heap* alloc() = 0;
+
+    // Free a heap item.
+    //
+    // HeapInterface::free can be called on objects returned from either
+    // HeapInterface::alloc or HeapInterface::get_heap_head.
+    virtual void free(t_heap* hptr) = 0;
```

Comment:
Add to comments:
This free puts the heap item on a heap-specific free list, to facilitate rapid re-use of heap elements, as the router rapidly creates and destroys heap items.

Reply:
That is the choice of the implementation of `HeapInterface`, so it is best to leave it off.